src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp

```diff
@@ -146,12 +146,7 @@ void AsynchronousReadIndirectBufferFromRemoteFS::prefetch(int64_t priority)
 
 void AsynchronousReadIndirectBufferFromRemoteFS::setReadUntilPosition(size_t position)
 {
-    /// Do not reinitialize internal state in case the new end of range is already included.
-    /// Actually it is likely that we will anyway reinitialize it as seek method is called after
-    /// changing end position, but seek avoiding feature might help to avoid reinitialization,
-    /// so this check is useful to save the prefetch for the time when we try to avoid seek by
-    /// reading and ignoring some data.
-    if (!read_until_position || position > *read_until_position)
+    if (!read_until_position || position != *read_until_position)
```

Comment:
This was making read_until_position useless on Parquet because it first reads metadata at the end, then everything else. Hope this check wasn't important in some other situation.

Reply:
changing this will make this case worse https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreePrefetchedReadPool.cpp#L230
e.g. there could be already prefetched data or data on the fly, but we will discard it for the first half of the tasks.
Though this case is not the most important, probably could be removed.

Reply:
I couldn't figure out how MergeTreePrefetchedReadPool task stealing is related to setReadUntilPosition(). AFAICT:
 * Each MergeTreeReadTask gets its own IMergeTreeReader, they're not shared or reused.
 * Each MergeTreeReaderWide (IMergeTreeReader) creates its own instances of MergeTreeReaderStream.
 * Each MergeTreeReaderStream creates its own ReadBuffer.
 * So ReadBuffers are not reused across MergeTreeReadTask, so it doesn't matter which task runs on which thread, as far as setReadUntilPosition() is concerned.

Which of these steps is incorrect?

Reply:
ReadBuffer are reused only in case the thread stole a task which reads from a part which this thread last read https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeThreadSelectProcessor.cpp#L58. 
But in the case I mentioned it does not matter if the part name is the same.
Prefetched read pool created a reader with async prefetch and it lies in MergeTreeReadTask as std::future. In the lines of code I attached, there is a tricky case when we want to steal half of the task of one thread. So the prefetch covers the first part of the task, we steal second part of the task. So then if we see in setReadUntilPosition that the right bound of the read task changed (it will be, as we split the task in two parts), we will reset the prefetch because it is always done when right bound changes.
Another point is that we could not reset prefetch in this case if we know that prefetch future is already valid and it will make this case better, currently it resets it unconditionally.

Reply:
> MergeTreeReadTask is never split into multiple tasks.  Its set of ranges is never modified. Work stealing moves tasks around but doesn't mutate/split/merge/create/destroy them.

It is not true in the case I was talking about, it was on the link of my first comment https://github.com/ClickHouse/ClickHouse/blob/35572321a14d617cfd110a48d8d3416615bd75c9/src/Storages/MergeTree/MergeTreePrefetchedReadPool.cpp#L230

> So, moving tasks between MergeTreePrefetchedReadPool::threads_tasks shouldn't affect setReadUntilPosition() calls much?

it affects because a reader with a prefetch can be created before the task is taken by some thread in getNewTask, and it could have a position already set to the boundaries of the task before it was split and some data according to initial task range was requested (though it is only data of 1 prefetch buffer size starting from the start of task range):

https://github.com/ClickHouse/ClickHouse/blob/35572321a14d617cfd110a48d8d3416615bd75c9/src/Storages/MergeTree/MergeTreePrefetchedReadPool.cpp#L106

so we have this case 

1. we created a prefetched task with [10, 30]
2. created a reader
3. decided to make a prefetch
4. seeked to 10
5. called setReadUntilPosition to 30
6. async prefetch is somewhere on the fly
7. then actual reading starts
8. some thread comes in getNewTask and decides to steal
9. the only option to steal is that prefetched task
10. he gets the second half of the task while the first half of the task remains, this first half will have a future with prefetched reader
11. when the next thread gets this the first half of splited task, it also does setReadUntilPosition on the reader it got from std::future from inside the task
12. now this setReadUntilPosition will be 20, though was 30, so it will reset the prefetch with the current changes


> I think I'll give up trying to understand how exactly merge tree reading works and just add that check for prefetch_future.valid()

yep, also an option.
Also as I said, this case with splitting the task is not an important case (I do not remember if there was perf difference from it), we can remove it as well from the code of prefetched read pool.

Reply:
> he gets the second half of the task while the first half of the task remains, this first half will have a future with prefetched reader

Where's the code that does this? The code after `/// Get second half of the tasks` only moves whole `MergeTreeReadTaskPtr`s between two `deque`s, doesn't mutate or create `MergeTreeReadTask`s.

Reply:
ah... indeed, I was thinking about the case which is not implemented, sorry for confusing you...